### PR TITLE
Deepcopy tests

### PR DESCRIPTION
**Pull request by soltysh**
@deads2k @sttts ptal

### CODE REVIEW

File: pkg/testing/deepcopy_test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		// switch to specialized. For now we're happy with the current serialization test.
+		fuzzer := fuzzer.FuzzerFor(genericfuzzer.Funcs, rand.NewSource(seed), codecs)
+		kinds := scheme.KnownTypes(gv)
+
+		for kind := range kinds {
+			if globalNonRoundTrippableTypes.Has(kind) {
+				continue
+			}
+			gvk := gv.WithKind(kind)
+			// FIXME: this is explicitly testing w/o protobuf which was failing
```
Comment (sttts) [inline pkg/testing/deepcopy_test.go]: for deepcopy this is enough. We only use serialization to get a 100% deepcopy with a 100% different code-path. So we don't care about the serialization format.

File: glide.yaml [hunk 2]
```diff
@@ -1,12 +1,13 @@
 package: github.com/openshift/api
 import:
 - package: k8s.io/gengo
-  repo:    git@github.com:openshift/kubernetes-gengo
```
Comment (deads2k) [inline glide.yaml]: I think we still want to use our fork.

Comment (deads2k) [reply glide.yaml]: Oh, reordered.  Uh ok.

File: glide.yaml [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
+  repo: git@github.com:openshift/kubernetes-gengo
 - package: k8s.io/code-generator
-  repo:    git@github.com:openshift/kubernetes-code-generator
   version: release-1.9.0-beta.1
+  repo: git@github.com:openshift/kubernetes-code-generator
 - package: k8s.io/apimachinery
   version: release-1.9
 - package: k8s.io/api
   version: release-1.9
+- package: github.com/json-iterator/go
```
Comment (deads2k) [inline glide.yaml]: I'm surprised to see this

File: pkg/testing/deepcopy_test.go (hunk 1 above)
Comment (deads2k) [reply pkg/testing/deepcopy_test.go]: > for deepcopy this is enough. We only use serialization to get a 100% deepcopy with a 100% different code-path. So we don't care about the serialization format.

We do at least a little bit.  I care why it was failing anyway.

File: pkg/testing/deepcopy_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+			roundtrip.RoundTripSpecificKindWithoutProtobuf(t, gvk, scheme, codecs, fuzzer, nil)
```
Comment (deads2k) [inline pkg/testing/deepcopy_test.go]: This is testing deep copy somehow?  It's not very explicit.  Perhaps renaming the function upstream or making something more targeted for deepcopy?

File: pkg/testing/deepcopy_test.go (hunk 1 above)
Comment (sttts) [reply pkg/testing/deepcopy_test.go]: > I care why it was failing anyway.

True. Just note our generic fuzzer. That's my guess: we put something in that the protobuf serialization does not like. But know what it is would be valuable.

File: pkg/testing/deepcopy_test.go (hunk 4 above)
Comment (sttts) [reply pkg/testing/deepcopy_test.go]: A comment would be better. It's kind of implicit:

1. `object := original.DeepCopyObject()`
2. `obj3 := Decode(Encode(object)`
3. `Fuzz(object)`
4. check `semantic.DeepEqual(original, obj3)`

We might want a negative test case as well to make sure it keeps testing the deepcopy.

**Comment by soltysh** [comment]
I've fixed deps and addressed all comments except for one (the negative test). But I think it can be taken care of as a followup.

**Comment by deads2k** [comment]
merged https://github.com/openshift/api/pull/21 as an alternate

**Pull request closed without merging**